services/index/src/helpers.js

```diff
@@ -1,111 +1,360 @@
+      ns = indexNamespace.parent + indexNamespace.name;
+    }
+    return {
+      namespace: ns,
+      name: indexNamespace.name,
+      expires: indexNamespace.expires.toJSON(),
+    };
+  },
+  // Call db.get_index_namespaces with named arguments.
+  // You can use this in two ways: with a handler or without a handler.
+  // In the latter case you'll get a list of up to 1000 entries and a
+  // continuation token.
+  // The response will be of the form { rows, continationToken }.
+  // If there are no namespaces to show, the response will have the
+  // `rows` field set to an empty array.
+  //
+  // If a handler is supplied, then it will invoke the handler
+  // on every item of the scan.
+  async getNamespaces(
+    db,
+    { parent, name },
+    {
+      query,
+      handler,
+    } = {},
+  ) {
+    assert(!handler || handler instanceof Function,
+      'If options.handler is given it must be a function');
+    const fetchResults = async (continuation) => {
+      let q = query;
 
-/** Regular expression for valid namespaces */
-exports.namespaceFormat = /^([a-zA-Z0-9_!~*'()%-]+\.)*[a-zA-Z0-9_!~*'()%-]+$/;
+      if (continuation) {
+        q.continuationToken = continuation;
+      }
+
+      const {continuationToken, rows} = await paginateResults({
+        query: q,
+        fetch: (size, offset) => db.fns.get_index_namespaces(
+          parent || parent === '' ? parent : null,
+          name || name === '' ? name : null,
+          size,
+          offset,
+        ),
+      });
+
+      const entries = rows.map(exports.namespaceUtils.fromDb);
+
+      return { rows: entries, continuationToken: continuationToken };
+    };
+
+    // Fetch results
+    let results = await fetchResults(query ? query.continuationToken : {});
+
+    // If we have a handler, then we have to handle the results
+    if (handler) {
+      const handleResults = async (res) => {
+        await Promise.all(res.rows.map((item) => handler.call(this, item)));
+
+        if (res.continuationToken) {
+          return await handleResults(await fetchResults(res.continuationToken));
+        }
+      };
+      results = await handleResults(results);
+    }
+    return results;
+  },
+  /** Create parent structure */
+  ensureNamespace(db, namespace, expires) {
+    // Stop recursion at root
+    if (namespace.length === 0) {
+      return Promise.resolve(null);
+    }
+
+    // Round to date to avoid updating all the time
+    expires = new Date(
+      expires.getFullYear(),
+      expires.getMonth(),
+      expires.getDate() + 1,
+      0, 0, 0, 0,
+    );
+
+    // Parse namespace
+    if (!(namespace instanceof Array)) {
+      namespace = namespace.split('.');
+    }
+    // Find parent and folder name
+    let name = namespace.pop() || '';
+    let parent = namespace.join('.');
+
+    // Load namespace, to check if it exists and if we should update expires
+    return db.fns.get_index_namespace(parent, name)
+      .then(async function(f) {
```

Comment:
Mixing `.then` and `async` is a little scary -- perhaps convert the whole thing to `async`?